docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+and a new `Handler` struct. This core API aims to:
+- be simpler,
+- more extensible,
+- more stable than the current framework,
+- enable deterministic events and queries, 
+- support event listeners and
+- [ADR 033: Protobuf-based Inter-Module Communication](./adr-033-protobuf-inter-module-comm.md) clients.
+
+## Context
+
+Historically modules have exposed their functionality to the framework via the `AppModule` and `AppModuleBasic`
+interfaces which have the following shortcomings:
+* both `AppModule` and `AppModuleBasic` need to be defined and registered which is counter-intuitive
+* apps need to implement the full interfaces, even parts they don't need (although there are workarounds for this),
+* interface methods depend heavily on unstable third party dependencies, in particular Tendermint,
+* legacy required methods have littered these interfaces for far too long
+
+In order to interact with the state machine, modules have needed to do a combination of these things:
+* get store keys from the app
+* call methods on `sdk.Context` which contains more or less the full set of capability available to modules.
+
+By isolating all the state machine functionality into `sdk.Context`, the set of functionalities available to
+modules are tightly coupled to this type. If there are changes to upstream dependencies (such as Tendermint)
+or new functionalities are desired (such as alternate store types), the changes need impact `sdk.Context` and all
+consumers of it (basically all modules). Also, all modules now receive `context.Context` and need to convert these
+to `sdk.Context`'s with a non-ergonomic unwrapping function.
+
+Any breaking changes to these interfaces, such as ones imposed by third-party dependencies like Tendermint, have the
+side effect of forcing all modules in the ecosystem to update in lock-step. This means it is almost impossible to have
+a version of the module which can be run with 2 or 3 different versions of the SDK or 2 or 3 different versions of
+another module. This lock-step coupling slows down overall development within the ecosystem and causes updates to
+components to be delayed longer than they would if things were more stable and loosely coupled.
+
+## Decision
+
+The `core` API proposes a set of core APIs that modules can rely on to interact with the state machine and expose their
+functionalities to it that are designed in a principled way such that:
+* tight coupling of dependencies and unrelated functionalities is minimized or eliminated
+* APIs can have long-term stability guarantees
+* the SDK framework is extensible in a safe and straightforward way
+
+The design principles of the core API are as follows:
+* everything that a module wants to interact with in the state machine is a service
+* all services coordinate state via `context.Context` and don't try to recreate the "bag of variables" approach of `sdk.Context`
```

Comment:
Just to understand this line: this "bag of variables" still exists, and will always exist with this new design, correct? It's just that we try our best to hide it to module developers.

As long as most keeper methods (e.g. msg server methods) have access to a `context.Context`, they can always peek into it to see all available variables, even those not exposed by said keeper's available services.

Reply:
Sort of, but there's no way to do a range operation over everything in a context. You need to hold a key object which can be non-forgeable.

Reply:
But you can still access `sdk.Context` via the unwrap method.

Is the plan to deprecate `sdk.Context`? If yes, it's worth mentioning this in the ADR, and start adding `// Deprecated` comments on it as soon as this approach is approved by the community.

Reply:
For now `sdk.Context` can still be accessed via unwrap. In my mind deprecating `sdk.Context` is somewhat orthogonal because that's a runtime impl detail, but I can say something here